src/client/components/DicePanel.tsx

```diff
@@ -1,57 +1,382 @@
 import {
-  faBalanceScale,
-  faBroom,
   faDiceD20,
-  faDragon,
-  faFireAlt,
-  faMagic,
+  faHandPaper,
+  faPlus,
   faShieldAlt,
 } from "@fortawesome/free-solid-svg-icons";
 import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
 import clsx from "clsx";
-import React, { useState } from "react";
+import React, { useState, useEffect } from "react";
+import {
+  playerAddDiceTemplateCategory,
+  playerUpdateDiceTemplateCategory,
+  playerDeleteDiceTemplateCategory,
+} from "../../shared/actions";
+import { DEFAULT_SYNC_TO_SERVER_DEBOUNCE_TIME } from "../../shared/constants";
+import {
+  categoryIcons,
+  characterStatNames,
+  fixedCategoryIcons,
+  iconMap,
+  RRDiceTemplatePart,
+  RRDiceTemplatePartDice,
+  skillMap,
+  skillNames,
+  userCategoryIcons,
+} from "../../shared/state";
+import { rrid } from "../../shared/util";
+import {
+  RRDiceTemplate,
+  RRDiceTemplateCategory,
+} from "../../shared/validation";
+import { useMyProps } from "../myself";
+import { useServerDispatch, useServerState } from "../state";
 import { DiceInput } from "./DiceInput";
 import { DiceInterface } from "./DiceInterface";
-import { DiceTemplates } from "./DiceTemplates";
-
-export const diceCategories = [
-  faFireAlt,
-  faBalanceScale,
-  faShieldAlt,
-  faMagic,
-  faDragon,
-  faBroom,
-];
+import { DiceTemplates, GeneratedDiceTemplates } from "./DiceTemplates";
+import { Popover } from "./Popover";
+import { Button } from "./ui/Button";
+import { SmartTextInput } from "./ui/TextInput";
 
 export const DicePanel = React.memo(function DicePanel() {
-  const [active, setActive] = useState(0);
+  const [active, setActive] = useState("Input");
+  const myself = useMyProps("diceTemplateCategories", "id", "mainCharacterId");
+  const dispatch = useServerDispatch();
+
+  useEffect(() => {
+    if (
+      active !== "Skills" &&
+      active !== "STs" &&
+      myself.diceTemplateCategories.find((cat) => cat.id === active) ===
+        undefined
+    ) {
+      setActive("Input");
+    }
+  }, [myself.diceTemplateCategories, active]);
+
+  function addTemplateCategory() {
+    const freeIcon = userCategoryIcons.find((userIcon) => {
+      return !myself.diceTemplateCategories
+        .map(({ icon }: { icon: typeof categoryIcons[number] }) => icon)
+        .includes(userIcon);
+    });
+
+    if (!freeIcon) return;
+
+    dispatch({
+      actions: [
+        playerAddDiceTemplateCategory({
+          id: myself.id,
+          category: {
+            id: rrid<RRDiceTemplateCategory>(),
+            icon: freeIcon,
+            categoryName: "",
+            templates: [],
+          },
+        }),
+      ],
+      optimisticKey: "diceTemplateCategories",
+      syncToServerThrottle: DEFAULT_SYNC_TO_SERVER_DEBOUNCE_TIME,
```

Comment:
```suggestion
      syncToServerThrottle: 0,
```
This could probably be 0, since it isn't something that is triggered repeatedly like the `onChange` event of a text input. There is no harm in sending this action to the server immediately.